core/src/hir/elision.rs

```diff
@@ -0,0 +1,409 @@
+//! argument and the lifetime of `s`, making us reject this method. Therefore, we
+//! have to be extremely careful when traversing lifetimes, and make sure that
+//! lifetimes of `Self` are lowered but _not_ considered for elision, while other
+//! lifetimes are lowered while also being considered for elision.
+//!
+//! # Lowering and Inference
+//!
+//! Lowering and elision inference is broken into three distinct stages:
+//! 1. Lowering the borrow in `&self` or `&mut self`, if there is one.
+//! 2. Lowering lifetimes of other params.
+//! 3. Lowering lifetimes of the output.
+//!
+//! Although each stage fundementally lowers lifetimes, they behave differently
+//! when lowering elided lifetimes. Naturally, this module represents each stage
+//! as a state in a state machine.
+//!
+//! The first state is represented by the [`SelfParamLifetimeLowerer`] type.
+//! Since there is either zero or one occurrences of `&self` or `&mut self`, it
+//! exposes the `.no_self_ref()` and `.lower_self_ref(lt)` methods respectively,
+//! which consume the `SelfParamLifetimeLowerer` and return the next state,
+//! [`ParamLifetimeLowerer`], as well as the lowered lifetime. The reason these
+//! are two distinct types is that the lifetime in `&self` and `&mut self` takes
+//! precedence over any other lifetimes in the input, so `.lower_self_ref(lt)`
+//! tells the next state that the candidate lifetime is already found, and to
+//! generate fresh anonymous lifetimes for any elided lifetimes.
+//!
+//! The second state is represented by the [`ParamLifetimeLowerer`] type.
+//! It implements a helper trait, [`LifetimeLowerer`], which abstracts the lowering
+//! of references and generic lifetimes. Internally, it wraps an [`ElisionSource`],
+//! which acts as a state machine for tracking candidate lifetimes to correspond
+//! to elision in the output. When a lifetime that's not in the type of the `self`
+//! argument or in the expanded generics of the `Self` type is visited, this
+//! state machine is potentially updated to another state. If the lifetime is
+//! anonymous, it's added to the internal list of nodes that go into the final
+//! [`LifetimeEnv`] after lowering. Once all the lifetimes in the input are
+//! lowered, the `into_return_ltl()` method is called to transition into the
+//! final state.
+//!
+//! The third and final state is represented by the [`ReturnLifetimeLowerer`] type.
+//! Similar to `ParamLifetimeLowerer`, it also implements the [`LifetimeLowerer`]
+//! helper trait. However, it differs from `ParamLifetimeLowerer` since instead
+//! of potentially updating the internal `ElisionSource` when visiting a lifetime,
+//! it instead reads from it when an elided lifetime occurs. Once all the output
+//! lifetimes are lowered, `.finish()` is called to return the finalized
+//! [`LifetimeEnv`].
+//!
+//! [Nomicon]: https://doc.rust-lang.org/nomicon/lifetime-elision.html
+
+use super::{
```

Comment:
nit: for future reference, if moving a large chunk of code prefer doing it in a separate commit so it's easier to review for actual changes

Reply:
(also better to do formatting in separate commits for the same reason)

Reply:
that makes a lot of sense, now I know